vnext/Microsoft.ReactNative/TurboModulesProvider.cpp

```diff
@@ -0,0 +1,169 @@
+    auto key = to_string(name);
+    VerifyElseCrash(m_methods.find(key) == m_methods.end());
+    m_methods.insert({key, {returnType, method}});
+  }
+
+  void AddSyncMethod(hstring const &name, SyncMethodDelegate const &method) noexcept {
+    VerifyElseCrash(false);
+  }
+
+ public:
+  std::unordered_map<std::string, TurboModuleMethodInfo> m_methods;
+};
+
+/*-------------------------------------------------------------------------------
+  TurboModuleImpl
+-------------------------------------------------------------------------------*/
+
+class TurboModuleImpl : public facebook::react::TurboModule {
+ public:
+  TurboModuleImpl(
+      const std::string &name,
+      std::shared_ptr<facebook::react::CallInvoker> jsInvoker,
+      ReactModuleProvider reactModuleProvider)
+      : facebook::react::TurboModule(name, jsInvoker), m_moduleBuilder(winrt::make<TurboModuleBuilder>()) {
+    providedModule = reactModuleProvider(m_moduleBuilder);
+  }
+
+  facebook::jsi::Value get(facebook::jsi::Runtime &runtime, const facebook::jsi::PropNameID &propName) override {
+    // it is not safe to assume that "runtime" never changes, so members are not cached here
+    auto tmb = m_moduleBuilder.as<TurboModuleBuilder>();
+    auto key = propName.utf8(runtime);
+    {
+      // try to find a Method
+      auto it = tmb->m_methods.find(key);
+      if (it != tmb->m_methods.end()) {
+        return facebook::jsi::Function::createFromHostFunction(
+            runtime,
+            propName,
+            0,
+            [&runtime, method = it->second](
+                facebook::jsi::Runtime &rt,
+                const facebook::jsi::Value &thisVal,
+                const facebook::jsi::Value *args,
+                size_t count) {
+              // prepare input arguments
+              facebook::jsi::Array argsArray(runtime, count);
+              for (size_t i = 0; i < count; i++) {
+                argsArray.setValueAtIndex(runtime, i, args[i]);
+              }
+              auto argReader = winrt::make<JsiReader>(runtime, facebook::jsi::Value(runtime, argsArray));
+
+              // prepare output value
+              auto argWriter = winrt::make<JsiWriter>(runtime);
+
+              // call the function
+              switch (method.ReturnType) {
+                case MethodReturnType::Void: {
+                  method.Method(argReader, argWriter, nullptr, nullptr);
+                  return facebook::jsi::Value::undefined();
+                }
+                case MethodReturnType::Promise: {
+                  return facebook::react::createPromiseAsJSIValue(
+                      runtime, [=](facebook::jsi::Runtime &runtime, std::shared_ptr<facebook::react::Promise> promise) {
+                        method.Method(
+                            argReader,
+                            argWriter,
+                            [promise, &runtime](const IJSValueWriter &writer) {
```

Comment:
>&runtime [](start = 38, length = 8)

If we pass runtime like this, then the method must be synchronous. Otherwise, it will crash.